Unable to build/test: lint `private_no_mangle_fns` has been removed

Running `cargo test --release --features=use-mock-routing` and/or `cargo build --release` yields:

```
   Compiling safe_core v0.32.0 ($HOME\safe_client_libs\safe_core)
error: lint `private_no_mangle_fns` has been removed: `no longer an warning, #[no_mangle] functions always exported`
  --> safe_core\src\lib.rs:93:5
   |
93 |     private_no_mangle_fns,
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> safe_core\src\lib.rs:83:5
   |
83 |     warnings
   |     ^^^^^^^^
   = note: #[forbid(renamed_and_removed_lints)] implied by #[forbid(warnings)]

error: lint `private_no_mangle_statics` has been removed: `no longer an warning, #[no_mangle] statics always exported`
  --> safe_core\src\lib.rs:94:5
   |
94 |     private_no_mangle_statics,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

error: Could not compile `safe_core`.

To learn more, run the command again with --verbose.
```

That's it, thanks! I was using 1.31.0 indeed.

Building now works, but testing gives some errors about `safe_app_jni` not being able to compile. (Might be due to Windows/Linux; haven't tried yet on Linux.)

```
   Compiling safe_app_jni v0.1.0 (file:///xxx/safe_client_libs/safe_app_jni)
error: linking with `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\link.exe` failed: exit code: 1120
```

With a lot of notes like this:

```
  = note: safe_authenticator.lib(keypair.obj) : warning LNK4099: PDB 'libsodium.pdb' was not found with 'safe_authenticator.lib(keypair.obj)' or at 'xxx\safe_client_libs\target\release\deps\libsodium.pdb'; linking object as if no debug info
```

That's it, thanks! I was using 1.31.0 indeed.

Building now works, but testing gives some errors about `safe_app_jni` not being able to compile. (Might be due to Windows/Linux; haven't tried yet on Linux.)

```
   Compiling safe_app_jni v0.1.0 (file:///xxx/safe_client_libs/safe_app_jni)
error: linking with `C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\link.exe` failed: exit code: 1120
```

With a lot of notes like this:

```
  = note: safe_authenticator.lib(keypair.obj) : warning LNK4099: PDB 'libsodium.pdb' was not found with 'safe_authenticator.lib(keypair.obj)' or at 'xxx\safe_client_libs\target\release\deps\libsodium.pdb'; linking object as if no debug info
```

--- a/safe_app/src/ffi/nfs.rs
+++ b/safe_app/src/ffi/nfs.rs
@@ -321,7 +321,7 @@ pub unsafe extern "C" fn file_write(
 /// Close is invoked only after all the data is completely written. The
 /// file is saved only when `close` is invoked.
 ///
-/// If the file was opened in any of the read modes, returns the modified
+/// If the file was opened in any of the write modes, returns the modified
 /// file structure as a result. If the file was opened in the read mode,
 /// returns the original file structure that was passed as an argument to
 /// `file_open`.
@@ -350,7 +350,7 @@ pub unsafe extern "C" fn file_close(
                     }).into_box()
                     .into()
             } else {
-                // The reader will be dropped automatically
+                // The reader will be dropped automatically.
                 o_cb(
                     user_data.0,
                     FFI_RESULT_OK,

--- a/safe_core/src/nfs/reader.rs
+++ b/safe_core/src/nfs/reader.rs
@@ -14,8 +14,8 @@ use self_encryption::SelfEncryptor;
 use self_encryption_storage::SelfEncryptionStorage;
 use utils::FutureExt;
 
-/// Reader is used to read contents of a File. It can read in chunks if the
-/// file happens to be very large.
+/// `Reader` is used to read contents of a `File`. It can read in chunks if the `File` happens to be
+/// very large.
 #[allow(dead_code)]
 pub struct Reader<C: Client> {
     client: C,
@@ -23,7 +23,7 @@ pub struct Reader<C: Client> {
 }
 
 impl<C: Client> Reader<C> {
-    /// Create a new instance of Reader.
+    /// Create a new instance of `Reader`.
     pub fn new(
         client: C,
         storage: SelfEncryptionStorage<C>,

--- a/safe_core/src/nfs/tests.rs
+++ b/safe_core/src/nfs/tests.rs
@@ -47,16 +47,19 @@ fn create_test_file_with_size(
                 Mode::Overwrite,
                 root.enc_key().cloned(),
             )
-        }).then(move |res| {
+        })
+        .then(move |res| {
             let writer = unwrap!(res);
 
             let bytes = vec![0u8; size];
             writer.write(&bytes).and_then(move |_| writer.close())
-        }).then(move |res| {
+        })
+        .then(move |res| {
             let file = unwrap!(res);
 
             file_helper::insert(c3, root2.clone(), "hello.txt", &file).map(move |_| (root2, file))
-        }).into_box()
+        })
+        .into_box()
 }
 
 fn create_test_file(client: &CoreClient) -> Box<NfsFuture<(MDataInfo, File)>> {
@@ -93,26 +96,31 @@ fn file_fetch_public_md() {
                     Mode::Overwrite,
                     root.enc_key().cloned(),
                 )
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let writer = unwrap!(res);
 
                 writer
                     .write(&[0u8; ORIG_SIZE])
                     .and_then(move |_| writer.close())
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let file = unwrap!(res);
 
                 file_helper::insert(c3, root2.clone(), "", &file).map(move |_| root2)
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let dir = unwrap!(res);
 
                 file_helper::fetch(c4, dir.clone(), "").map(move |(_version, file)| (dir, file))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (dir, file) = unwrap!(res);
 
                 file_helper::read(c5, &file, dir.enc_key().cloned())
                     .map(move |reader| (reader, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (reader, dir) = unwrap!(res);
                 let size = reader.size();
                 println!("reading {} bytes", size);
@@ -120,18 +128,21 @@ fn file_fetch_public_md() {
                     assert_eq!(data, vec![0u8; ORIG_SIZE]);
                     dir
                 })
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let dir = unwrap!(res);
 
                 std::thread::sleep(std::time::Duration::new(3, 0));
 
                 file_helper::fetch(c6, dir.clone(), "").map(move |(_version, file)| (dir, file))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (dir, file) = unwrap!(res);
 
                 file_helper::read(c7, &file, dir.enc_key().cloned())
                     .map(move |reader| (reader, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (reader, _dir) = unwrap!(res);
                 let size = reader.size();
                 println!("reading {} bytes", size);
@@ -156,7 +167,8 @@ fn file_read() {
 
                 file_helper::read(c2, &file, dir.enc_key().cloned())
                     .map(move |reader| (reader, file, creation_time))
-            }).then(|res| {
+            })
+            .then(|res| {
                 let (reader, file, creation_time) = unwrap!(res);
                 let size = reader.size();
                 println!("reading {} bytes", size);
@@ -166,7 +178,8 @@ fn file_read() {
                 assert!(creation_time <= *file.modified_time());
 
                 result
-            }).map(move |data| {
+            })
+            .map(move |data| {
                 assert_eq!(data, vec![0u8; ORIG_SIZE]);
             })
     });
@@ -185,7 +198,8 @@ fn file_read_chunks() {
                 let (dir, file) = unwrap!(res);
 
                 file_helper::read(c2, &file, dir.enc_key().cloned())
-            }).then(|res| {
+            })
+            .then(|res| {
                 let reader = unwrap!(res);
                 let size = reader.size();
                 assert_eq!(size, ORIG_SIZE as u64);
@@ -216,7 +230,8 @@ fn file_read_chunks() {
                             }
                         })
                     },
-                ).then(
+                )
+                .then(
                     move |res: Result<(Reader<CoreClient>, u64, Vec<u8>), NfsError>| {
                         let (reader, size_read, result) = unwrap!(res);
 
@@ -227,13 +242,15 @@ fn file_read_chunks() {
                         println!("reading 0 bytes");
                         reader.read(size, 0).map(move |data| (reader, size, data))
                     },
-                ).then(|res| {
+                )
+                .then(|res| {
                     let (reader, size, data) = unwrap!(res);
                     assert_eq!(data, Vec::<u8>::new());
 
                     // Read past the end of the file, expect an error
                     reader.read(size, 1)
-                }).then(|res| -> Result<_, CoreError> {
+                })
+                .then(|res| -> Result<_, CoreError> {
                     match res {
                         Ok(_) => {
                             // We expect an error in this case
@@ -265,7 +282,8 @@ fn file_write_chunks() {
 
                 file_helper::write(c2, file, Mode::Overwrite, dir.enc_key().cloned())
                     .map(move |writer| (writer, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, dir) = unwrap!(res);
 
                 let size_written = 0;
@@ -289,8 +307,10 @@ fn file_write_chunks() {
                                 Ok(Loop::Break(writer))
                             }
                         })
-                }).map(move |writer| (writer, dir))
-            }).then(
+                })
+                .map(move |writer| (writer, dir))
+            })
+            .then(
                 move |res: Result<(Writer<CoreClient>, MDataInfo), NfsError>| {
                     let (writer, dir) = unwrap!(res);
                     // Write 0 bytes, should succeed
@@ -298,16 +318,19 @@ fn file_write_chunks() {
                         .write(&content[GOAL_SIZE..GOAL_SIZE])
                         .map(move |_| (writer, dir))
                 },
-            ).then(move |res| {
+            )
+            .then(move |res| {
                 let (writer, dir) = unwrap!(res);
                 writer.close().map(move |file| (file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 // Updating file - append
                 let (file, dir) = unwrap!(res);
 
                 file_helper::write(c3, file, Mode::Append, dir.enc_key().cloned())
                     .map(move |writer| (writer, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, dir) = unwrap!(res);
 
                 let size_written = 0;
@@ -331,8 +354,10 @@ fn file_write_chunks() {
                                 Ok(Loop::Break(writer))
                             }
                         })
-                }).map(move |writer| (writer, dir))
-            }).then(
+                })
+                .map(move |writer| (writer, dir))
+            })
+            .then(
                 move |res: Result<(Writer<CoreClient>, MDataInfo), NfsError>| {
                     let (writer, dir) = unwrap!(res);
                     // Write 0 bytes, should succeed
@@ -340,20 +365,24 @@ fn file_write_chunks() {
                         .write(&content[GOAL_SIZE..GOAL_SIZE])
                         .map(move |_| (writer, dir))
                 },
-            ).then(move |res| {
+            )
+            .then(move |res| {
                 let (writer, dir) = unwrap!(res);
                 writer.close().map(move |file| (file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (file, dir) = unwrap!(res);
 
                 file_helper::read(c4, &file, dir.enc_key().cloned())
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let reader = unwrap!(res);
                 let size = reader.size();
 
                 assert_eq!(size, 2 * GOAL_SIZE as u64);
                 reader.read(0, size)
-            }).map(move |data| {
+            })
+            .map(move |data| {
                 assert_eq!(data, vec![0u8; 2 * GOAL_SIZE]);
             })
     })
@@ -377,21 +406,25 @@ fn file_update_overwrite() {
 
                 file_helper::write(c2, file, Mode::Overwrite, dir.enc_key().cloned())
                     .map(move |writer| (writer, dir, creation_time))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, dir, creation_time) = unwrap!(res);
                 writer
                     .write(&[1u8; NEW_SIZE])
                     .and_then(move |_| writer.close())
                     .map(move |file| (file, dir, creation_time))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (file, dir, creation_time) = unwrap!(res);
                 file_helper::update(c3, dir.clone(), "hello.txt", &file, Version::Custom(1))
                     .map(move |_| (dir, creation_time))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (dir, creation_time) = unwrap!(res);
                 file_helper::fetch(c4, dir.clone(), "hello.txt")
                     .map(move |(_version, file)| (dir, file, creation_time))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (dir, file, creation_time) = unwrap!(res);
 
                 // Check file timestamps
@@ -399,12 +432,14 @@ fn file_update_overwrite() {
                 assert!(creation_time <= *file.modified_time());
 
                 file_helper::read(c5, &file, dir.enc_key().cloned())
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let reader = unwrap!(res);
                 let size = reader.size();
                 println!("reading {} bytes", size);
                 reader.read(0, size)
-            }).map(move |data| {
+            })
+            .map(move |data| {
                 assert_eq!(data, vec![1u8; NEW_SIZE]);
             })
     });
@@ -430,21 +465,25 @@ fn file_update_append() {
                         // Updating file - append
                         file_helper::write(c2, file, Mode::Append, dir.enc_key().cloned())
                             .map(move |writer| (dir, writer))
-                    }).then(move |res| {
+                    })
+                    .then(move |res| {
                         let (dir, writer) = unwrap!(res);
                         writer
                             .write(&[2u8; APPEND_SIZE])
                             .and_then(move |_| writer.close())
                             .map(move |file| (dir, file))
-                    }).then(move |res| {
+                    })
+                    .then(move |res| {
                         let (dir, file) = unwrap!(res);
                         file_helper::read(c3, &file, dir.enc_key().cloned())
-                    }).then(move |res| {
+                    })
+                    .then(move |res| {
                         let reader = unwrap!(res);
                         let size = reader.size();
                         println!("reading {} bytes", size);
                         reader.read(0, size)
-                    }).map(move |data| {
+                    })
+                    .map(move |data| {
                         assert_eq!(data.len(), size + APPEND_SIZE);
                         assert_eq!(data[0..size].to_owned(), vec![0u8; size]);
                         assert_eq!(&data[size..], [2u8; APPEND_SIZE]);
@@ -473,11 +512,13 @@ fn file_update_metadata() {
                         dir
                     },
                 )
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let dir = unwrap!(res);
 
                 file_helper::fetch(c3.clone(), dir, "hello.txt")
-            }).map(move |(_version, file)| {
+            })
+            .map(move |(_version, file)| {
                 assert_eq!(*file.user_metadata(), [12u8; 10][..]);
             })
     });
@@ -497,10 +538,12 @@ fn file_delete() {
                         dir
                     },
                 )
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let dir = unwrap!(res);
                 file_helper::fetch(c3.clone(), dir, "hello.txt")
-            }).then(move |res| -> Result<_, CoreError> {
+            })
+            .then(move |res| -> Result<_, CoreError> {
                 match res {
                     Ok(_) => {
                         // We expect an error in this case
@@ -528,36 +571,43 @@ fn file_delete_then_add() {
                 let (dir, file) = unwrap!(res);
                 file_helper::delete(c2, dir.clone(), "hello.txt", Version::Custom(1))
                     .map(move |_| (dir, file))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (dir, file) = unwrap!(res);
 
                 file_helper::write(c3, file, Mode::Overwrite, dir.enc_key().cloned())
                     .map(move |writer| (writer, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, dir) = unwrap!(res);
 
                 writer
                     .write(&[1u8; NEW_SIZE])
                     .and_then(move |_| writer.close())
                     .map(move |file| (file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (file, dir) = unwrap!(res);
                 file_helper::update(c4, dir.clone(), "hello.txt", &file, Version::GetNext)
                     .map(move |_| dir)
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let dir = unwrap!(res);
                 file_helper::fetch(c5, dir.clone(), "hello.txt")
                     .map(move |(version, file)| (version, file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (version, file, dir) = unwrap!(res);
                 assert_eq!(version, 2);
                 file_helper::read(c6, &file, dir.enc_key().cloned())
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let reader = unwrap!(res);
                 let size = reader.size();
                 println!("reading {} bytes", size);
                 reader.read(0, size)
-            }).map(move |data| {
+            })
+            .map(move |data| {
                 assert_eq!(data, vec![1u8; NEW_SIZE]);
             })
     });
@@ -578,31 +628,129 @@ fn file_open_close() {
                 // Open the file for reading
                 file_helper::read(c2, &file, dir.enc_key().cloned())
                     .map(move |reader| (reader, file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 // The reader should get dropped implicitly
                 let (_reader, file, dir) = unwrap!(res);
                 // Open the file for writing
                 file_helper::write(c3, file.clone(), Mode::Overwrite, dir.enc_key().cloned())
                     .map(move |writer| (writer, file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, file, dir) = unwrap!(res);
                 // Close the file
                 let _ = writer.close();
                 // Open the file for appending
                 file_helper::write(c4, file.clone(), Mode::Append, dir.enc_key().cloned())
                     .map(move |writer| (writer, file, dir))
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let (writer, file, dir) = unwrap!(res);
                 // Close the file
                 let _ = writer.close();
                 // Open the file for reading, ensure it has original contents
                 file_helper::read(c5, &file, dir.enc_key().cloned())
-            }).then(move |res| {
+            })
+            .then(move |res| {
                 let reader = unwrap!(res);
                 let size = reader.size();
                 reader.read(0, size)
-            }).map(move |data| {
+            })
+            .map(move |data| {
                 assert_eq!(data, vec![0u8; ORIG_SIZE]);
+            })
+    });
+}
+
+// Test opening the same file multiple times concurrently.
+#[test]
+fn file_open_concurrent() {
+    random_client(|client| {
+        let c2 = client.clone();
+        let c3 = client.clone();
+        let c4 = client.clone();
+        let c5 = client.clone();
+        let c6 = client.clone();
+
+        create_test_file(client)
+            .then(move |res| {
+                let (dir, file) = unwrap!(res);
+
+                // Open the first writer.
+                file_helper::write(c2, file.clone(), Mode::Overwrite, dir.enc_key().cloned())
+                    .map(move |writer1| (writer1, file, dir))
+            })
+            .then(move |res| {
+                let (writer1, file, dir) = unwrap!(res);
+
+                // Open the second writer concurrently.
+                file_helper::write(c3, file.clone(), Mode::Overwrite, dir.enc_key().cloned())
+                    .map(move |writer2| (writer1, writer2, file, dir))
+            })
+            .then(move |res| {
+                let (writer1, writer2, file, dir) = unwrap!(res);
+
+                // Open a reader concurrently.
+                file_helper::read(c4, &file, dir.enc_key().cloned())
+                    .map(move |reader| (writer1, writer2, reader, file, dir))
+            })
+            .then(move |res| {
+                let (writer1, writer2, reader, file, dir) = unwrap!(res);
+
+                // Write with the first writer.
+                writer1
+                    .write(&[1u8; NEW_SIZE])
+                    .and_then(move |_| writer1.close())
+                    .map(move |_| (writer2, reader, file, dir))
+            })
+            .then(move |res| {
+                let (writer2, reader, file, dir) = unwrap!(res);
+
+                // Write with the second writer.
+                writer2
+                    .write(&[2u8; NEW_SIZE])
+                    .and_then(move |_| writer2.close())
+                    .map(move |file2| (reader, file, file2, dir))
+            })
+            .then(move |res| {
+                let (reader, file, file2, dir) = unwrap!(res);
+
+                // Read with the reader, it should have neither of the written changes.
+                let size = reader.size();
+                reader.read(0, size).map(move |data| {
+                    assert_eq!(data, vec![0u8; ORIG_SIZE]);
+                    (file, file2, dir)
+                })
+            })
+            .then(move |res| {
+                let (file, file2, dir) = unwrap!(res);
+
+                // Open the original file for reading again, it should be unchanged.
+                file_helper::read(c5, &file, dir.enc_key().cloned())
+                    .map(|reader| (reader, file2, dir))
+            })
+            .then(move |res| {
+                let (reader, file2, dir) = unwrap!(res);
+
+                let size = reader.size();
+                reader.read(0, size).map(move |data| {
+                    assert_eq!(data, vec![0u8; ORIG_SIZE]);
+                    (file2, dir)
+                })
+            })
+            .then(move |res| {
+                let (file2, dir) = unwrap!(res);
+
+                // Open the file written by writer2.
+                file_helper::read(c6, &file2, dir.enc_key().cloned())
+            })
+            .then(move |res| {
+                let reader = unwrap!(res);
+
+                let size = reader.size();
+                reader
+                    .read(0, size)
+                    .map(move |data| assert_eq!(data, vec![2u8; NEW_SIZE]))
             })
     });
 }
@@ -629,10 +777,12 @@ fn encryption() {
             File::new(Vec::new()),
             Mode::Overwrite,
             Some(key.clone()),
-        ).then(move |res| {
+        )
+        .then(move |res| {
             let writer = unwrap!(res);
             writer.write(&content).and_then(move |_| writer.close())
-        }).then(move |res| {
+        })
+        .then(move |res| {
             // Attempt to read without an encryption key fails.
             let file = unwrap!(res);
             file_helper::read(c2, &file, None)
@@ -641,7 +791,8 @@ fn encryption() {
                     // TODO: assert the error is of the expected variant.
                     Ok(file)
                 })
-        }).then(move |res| {
+        })
+        .then(move |res| {
             // Attempt to read using incorrect encryption key fails.
             let file = unwrap!(res);
             file_helper::read(c3, &file, Some(wrong_key))
@@ -650,15 +801,18 @@ fn encryption() {
                     NfsError::CoreError(CoreError::SymmetricDecipherFailure) => Ok(file),
                     error => Err(error),
                 })
-        }).then(move |res| {
+        })
+        .then(move |res| {
             // Attempt to read using original encryption key succeeds.
             let file = unwrap!(res);
             file_helper::read(c4, &file, Some(key))
-        }).then(move |res| {
+        })
+        .then(move |res| {
             let reader = unwrap!(res);
             let size = reader.size();
             reader.read(0, size)
-        }).then(move |res| -> Result<_, NfsError> {
+        })
+        .then(move |res| -> Result<_, NfsError> {
             let retrieved_content = unwrap!(res);
             assert_eq!(retrieved_content, content2);
             Ok(())

--- a/safe_core/src/nfs/writer.rs
+++ b/safe_core/src/nfs/writer.rs
@@ -15,17 +15,17 @@ use self_encryption::SequentialEncryptor;
 use self_encryption_storage::SelfEncryptionStorage;
 use utils::FutureExt;
 
-/// Mode of the writer
+/// Mode of the writer.
 #[derive(Clone, Copy, Debug)]
 pub enum Mode {
-    /// Will create new data
+    /// Will create new data.
     Overwrite,
-    /// Will append content to the existing data
+    /// Will append content to the existing data.
     Append,
 }
 
 /// Writer is used to write contents to a File and especially in chunks if the
-/// file happens to be too large
+/// file happens to be too large.
 pub struct Writer<C: Client> {
     client: C,
     file: File,
@@ -34,7 +34,7 @@ pub struct Writer<C: Client> {
 }
 
 impl<C: Client> Writer<C> {
-    /// Create new instance of Writer
+    /// Create new instance of Writer.
     pub fn new(
         client: &C,
         storage: SelfEncryptionStorage<C>,
@@ -60,7 +60,7 @@ impl<C: Client> Writer<C> {
         .into_box()
     }
 
-    /// Data of a file/blob can be written in smaller chunks
+    /// Data of a file/blob can be written in smaller chunks.
     pub fn write(&self, data: &[u8]) -> Box<NfsFuture<()>> {
         trace!(
             "Writer writing file data of size {} into self-encryptor.",
@@ -72,9 +72,9 @@ impl<C: Client> Writer<C> {
             .into_box()
     }
 
-    /// close is invoked only after all the data is completely written. The
-    /// file/blob is saved only when the close is invoked. Returns the final
-    /// `File` with the data_map stored on the network.
+    /// close() should be invoked only after all the data is completely written. The file/blob is
+    /// saved only when close() is invoked. Returns the final `File` with the data_map stored on the
+    /// network.
     pub fn close(self) -> Box<NfsFuture<File>> {
         trace!("Writer induced self-encryptor close.");
 

